binding-hs/src/Client.hs

```diff
@@ -0,0 +1,192 @@
+{-# LANGUAGE DataKinds             #-}
+{-# LANGUAGE DeriveGeneric         #-}
+{-# LANGUAGE DuplicateRecordFields #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE OverloadedStrings     #-}
+{-# LANGUAGE TypeOperators         #-}
+
+module Client
+  ( envCreate
+  , envListAll
+  , envReset
+  , envStep
+  , envActionSpaceInfo
+  , envActionSpaceSample
+  , envActionSpaceContains
+  , envObservationSpaceInfo
+  , envMonitorStart
+  , envMonitorClose
+  , envClose
+  , upload
+  , shutdownServer
+  , EnvID (..)
+  , InstID (..)
+  , Environment (..)
+  , Observation (..)
+  , Step (..)
+  , Outcome (..)
+  , Info (..)
+  , Action (..)
+  , Monitor (..)
+  , Config (..)
+  ) where
+
+import           Data.Aeson          (FromJSON, Object (..), ToJSON)
+import qualified Data.HashMap.Strict as HM (HashMap (..))
+import           Data.Proxy          (Proxy (..))
+import qualified Data.Text           as T (Text)
+import           GHC.Generics
+import           Network.HTTP.Client (Manager, defaultManagerSettings,
+                                      newManager)
+import           Servant.API
+import           Servant.Client      (BaseUrl (..), ClientM, client)
+import           Servant.HTML.Lucid  (HTML)
+
+--------------------------------------------------------------------------------
+-- | Data types and instances
+
+newtype EnvID = EnvID { env_id :: T.Text }
+  deriving Generic
+
+instance ToJSON EnvID
+
+newtype InstID = InstID { instance_id :: T.Text }
+  deriving (Eq, Show, Generic)
+
+instance ToJSON InstID
+instance FromJSON InstID
+
+newtype Environment = Environment { all_envs :: HM.HashMap T.Text T.Text }
+  deriving (Eq, Show, Generic)
+
+instance ToJSON Environment
+instance FromJSON Environment
+
+newtype Observation = Observation { observation :: [Double] }
+  deriving (Eq, Show, Generic)
+
+instance ToJSON Observation
+instance FromJSON Observation
+
+data Step = Step
+  { action :: !Int
+  , render :: !Bool
+  } deriving Generic
+
+instance ToJSON Step
+instance FromJSON Step
+
+data Outcome = Outcome
+  { observation :: ![Double]
```

Comment:
I used Aeson.Value instead. Object is a Hashmap Text Value and observation is indeed a JSON object where the key is "observation". We already have the key part through the observation record field name so we only need a Value. Nonetheless, this might not be the best solution. A simple Aeson.Array might be sufficient.